### DISCUSSION TITLE
Is it possible to sort windows by activity or check which one is the active window?

### DISCUSSION THREAD

**Discussion by Bluey26** [labwc/labwc#1783] [General]
Hello.

I just have switched to labwc (lxqt-labwc) a few hours ago and i was wondering if its possible, in some way, to sort the windows by their activity, with the objective of selecting with a command line script, the last active window.
In Xorg i have been using something as this to check which one is the active window, and then select the N-1 (the window prior to the last one) window and focus it.

This part checks if pcmanfm-desktop0 is the active window

`$(xprop -id $(xprop -root _NET_ACTIVE_WINDOW | cut -d ' ' -f 5) WM_NAME | sed -e 's/.*"\(.*\)".*/\1/') == "pcmanfm-desktop0"`

This part changes the focused window to the N-1 active window:

`xprop -root|grep "^_NET_CLIENT_LIST_STACKING" | tr "," " " | wmctrl -ia $(awk '{print $(NF-1)}'`

This seems to work under Xorg for my use case(minimize all windows and then only show the N-1 window focused), but as Wayland does the things in a diferent way this does not work.

I have been testing commands i have found around in the internet and none of them seems to provide a way to do this any of this two things (to check the current window or to select the N-1 active window):

`wlrctl window list`

`lswt`

I have also been searching for alternatives for those xorg commands, but they do not have this kind of functions:
`ydotool` , `dotool `, `wlprop`  seem to be wayland orientated but they don't cover this usage.

This post is in here because i thought you may know what i should do in this case or what alternative could be present to do this. Everything seems to indicate that this is not possible in wayland, at least in this moment, so you can also confirm me this.

Thanks in advance.

I am using `labwc 0.7.1-152-gabd8e971` (pulled from the Archlinux AUR `labwc-git` package)

**Comment by droc12345** [comment]
Have you looked at the built-in window switcher? 
_man 5 labwc-config_ for details on using it.

The window switcher would allow you to do a lot of what you want, though it doesn't sort, it does keep the windows in reverse used order.

wlrctl does do a lot of what you want, but it's not documentated, so you have to read the source.

**Comment by johanmalm** [comment]
I suspect the only clean way is to write a long running client that listens to the foreign-toplevel protocol (like a panel, but without the graphical stuff).

**Comment by Bluey26** [comment]
Thank you for your answers!

Yes, i imagined this was not an easy thing, since wayland is in early adoption in most DEs.

I will try to see what i can obtain with `wlrctl`, i already have an command-line to minimize everything, which is neither documented:

`wlrctl toplevel minimize state:inactive && wlrctl toplevel minimize state:active`  which was taken from a lxqt page.

Thanks also for moving the issue, i dont know why i opened it in issues lol

**Reply by davidphilipbarr** [reply]
If you run lswt -j you can get the active window from there, I have a lousy pipe menu that identifies active and minimised (and theoretically full screen) windows.

**Comment by davidphilipbarr** [comment]
I've just noticed that lswt has made it easier to get the window state, but also I wonder if the watch switch could be modified to provide state info, which would make it easier to reference previous focused windows, just a thought.

**Comment by Bluey26** [comment]
> If you run lswt -j you can get the active window from there, I have a lousy pipe menu that identifies active and minimised (and theoretically full screen) windows.

Yes i just saw your code lines and looks awesome, the problem with my use case is that the program that i used (pcmanfm-desktop0 ) which is the file manager and the 'desktop module' is not listed in there so it seems that i cannot use that unless i change something( i have to see what).

The activity side is the hard one, to 'remember' the order in which the windows were closed/opened as `_NET_CLIENT_LIST_STACKING` seems to do.

**Comment by droc12345** [comment]
What are you're trying to do? It seems as if you're trying to make a type of window switcher.

There's one built in, and a single "alt tab" would change focus to the last window.
The windows are listed in reverse order, newest to oldest and you can select among them.
"alt tab" twice would select the window previous to the last one.

The only drawback, if you want to call it that, is that the window switcher is keyboard driven not mouse.

**Comment by Bluey26** [comment]
> What are you're trying to do? It seems as if you're trying to make a type of window switcher.
> 
> There's one built in, and a single "alt tab" would change focus to the last window. The windows are listed in reverse order, newest to oldest and you can select among them. "alt tab" twice would select the window previous to the last one.
> 
> The only drawback, if you want to call it that, is that the window switcher is keyboard driven not mouse.

My idea is the following: There is a showdesktop plugin in LXQt, but at the moment it does not work, WIP.
That does the following: you click the widget icon and all windows get unfocused.You click it again and all windows get activated and focus all the windows that were present in the screen when you clicked the icon the first time.

Since i came from openbox, i have had an issue with my conky panel (the issue is not present in labwc) and managed to use the script i described in https://github.com/lxqt/lxqt/issues/107#issuecomment-1824653789 to show conky even when i minimized all other windows (video also showing this).

This script does the following: It unfocus all the windows on click, and if the current focused window is `pcmanfm-desktop0`(the desktop, no windows), it will bring in focus the last focused window (and not all the windows). This allows me to only see what i would like to see in a fast way, instead of seeing all the windows focused in the screen (shown in the video, the stardict part).

Since lxqt started to bring wayland support, i moved with it to wayland and to labwc which resembles the most to openbox, and wanted to keep this function too.

PS: window switcher works fine in lxqt, its just that i do not want to cycle through all the windows to find what i look for, but just do this faster with the mouse.

**Reply by stefonarch** [reply]
>the 'desktop module' is not listed in there 

It is no toplevel window anymore but a background layer surface. See also
https://github.com/labwc/labwc/issues/455

Once implemented having a working showdesktop button (with customcommand or quicklauch which uses `wtype` )  is a piece of cake.

**Comment by droc12345** [comment]
I would think that a show desktop function would exist inside of lxqt, not labwc, since labwc has no notion of "desktop"

And implementing something likes this in labwc highlights the differences between good old X and wayland,
they don't always do things the same way, or even look at them in the same way.

**Comment by droc12345** [comment]
For further thought. If I show desktop in openbox, then every window goes away, including panels, but not the background (if set) as that's part of the root window. 

In wayland, to show the "desktop" along with the panels, do I also hide the wallpaper program, as it's a running app, 
not just some image written to a "root window".

You could use something like wlrctl to minimize all but the active window, but do you really care about minimizing 
windows on some other workspace (not visible) and neither lswt or wlrctl will show things like a panel or wallpaper, 
to be able to manipulate it.

**Reply by stefonarch** [reply]
>I would think that a show desktop function would exist inside of lxqt, not labwc, since labwc has no notion of "desktop"

LXQt has no knowledge about which windows are where - the WM has it and on X11 there are functions to access the WM. On wayland the showdesktop plugin works on kwin_wayland only atm because there's specific code, wayfire provides a shortcut which can be used.

>For further thought. If I show desktop in openbox, then every window goes away, including panels, but not the background (if set) as that's part of the root window.

That is an openbox session afaik. Using it with LXQt it acts as every showdesktop action, leaving the panels and the desktop window in place.
